### PR TITLE
Enables fastboot for ledger-tool

### PR DESCRIPTION
**Pull request by brooksprumo**
#### Problem

When using ledger-tool to perform an accounts-related work, it can be very slow due to always having to extract a snapshot archive each time, especially if we already have all the accounts state on disk and could use that instead.

#### Summary of Changes

Enable fastboot for ledger-tool by adding a new flag, `--boot-from-local-state`.

Specifically for all the subcommands that load a bank *except* `create-snapshot`. We still need to do some more work there w.r.t. plumbing accounts hashes and starting snapshot hashes. So for the sake of using fastboot for as much as possible, we do not allow generating snapshots with fastboot enabled.

#### Extra Testing

I spun up a validator on mnb to generate some snapshot archives and bank snapshots, then stopped the validator and ran `ledger-tool verify --boot-from-local-state` and confirmed the newer bank snapshot was booted from, instead of the older snapshot archive.

| boot from | unarchive time | bank rebuild time | clean time | shrink time | total time until processing ledger begins |
| --- | ---: | ---: | ---: | ---: | ---: |
| snapshot archives | 119 s | 88 s | 31 s | 131 s | 549 s |
| local state | 0 s | 49 s | 0 s | 0 s | 75 s |

(Probably will be interesting to see why rebuilding the bank takes ~50 seconds, but I haven't looked into that part yet.)

Yes, this is not exhaustive testing. Let me know if there's more tests I should run! I figure this is OK since:
1. This is only in master, at the very beginning of v1.17, so we have ample time to find and fix issues before a release
2. This is only in ledger-tool, so validators are not impacted (and not in all ledger-tool subcommands, either)
3. This is currently a hidden flag that users have to opt into
4. Fastboot can have large time savings in ledger-tool-heavy debug/verify cycles, especially with large accounts, such as what's found on the kin-sim cluster

### CODE REVIEW

**Comment by steviez** [comment]
You mentioned spinning up a validator to get your initial account set state. Another common workflow I'd say is grabbing a snapshot + blocks from some other source (copy from other node / pull from Google / etc), and then doing repeated verify runs locally. Am I accurate in believing that I could run `solana-ledger-tool verify` once without fast-boot which would unpack the snapshot, and then for subsequent `verify` runs, I'd be able use `--fast-boot` ?

**Comment by apfitzge** [comment]
> since by default ledger-tool will extract the accounts and snapshots into different directories than what --boot-from-local-state requires.

In my branch I had also done what you did here, and only support primary (validator generated) state.
While reviewing, had a thought and I wonder what you're thoughts are. Why not make `--boot-from-local-state` an enum:
```rust
#[derive(Default)]
enum BootFromLocalState {
    /// Boot from state generated by running a validator
    #[default]
    Validator,
    /// Boot from state generated by running ledger-tool
    LedgerTool,
}
```

I think we can make clap do something where if we just do `--boot-from-local-state` it defaults to enums default, or user can specify `--boot-from-local-state <validator|ledger-tool>`.

**Comment by apfitzge** [comment]
> Is this so that fastboot would work with the accounts.ledger-tool and snapshot.ledger-tool directories?

Yeah for working with the `.ledger-tool` suffixed-dirs. Got it, I figured there would probably be some problem with doing it.

> If we can inflate the snapshot archive into the supported directory structure that fastboot requires, then we'll be able to unify the load-a-snapshot code. The code would then: (1) extract a snapshot archive, then (2) load from snapshot directory. And fastboot would skip (1).

This definitely seems like the right path forward. 

Maybe not the right place to discuss, but a few considerations about the extraction of archives in ledger-tool.
Do we want to extract into the `.ledger-tool` dirs? If we do, then we're non-destructive with the validator snapshots. But that also means we will have multiple hardlink dirs... hmmm

**Review by apfitzge — DISMISSED** [review]
Looks good to me. Adds an additional option, which is currently hidden, and does not affect the normal boot path.

File: ledger/src/blockstore_processor.rs [hunk 1]
```diff
@@ -632,6 +632,7 @@ pub struct ProcessOptions {
     /// true if after processing the contents of the blockstore at startup, we should run an accounts hash calc
     /// This is useful for debugging.
     pub run_final_accounts_hash_calc: bool,
+    pub boot_from_local_state: bool,
```
Comment (jeffwashington) [inline ledger/src/blockstore_processor.rs:635]: I suggest a comment here for the next weary traveller. This is a new feature at the moment and will always be conditional on us having correctly saved local state previously on this validator. So, this bifurcated startup process and code path will always exist.

File: ledger/src/blockstore_processor.rs [hunk 2]
```diff
@@ -632,6 +632,7 @@ pub struct ProcessOptions {
     /// true if after processing the contents of the blockstore at startup, we should run an accounts hash calc
     /// This is useful for debugging.
     pub run_final_accounts_hash_calc: bool,
+    ///Enable booting from state already on disk, instead of extracting if from a snapshot archive.
```
Comment (jeffwashington) [inline ledger/src/blockstore_processor.rs:635]: space after /// here?

File: ledger/src/blockstore_processor.rs (hunk 1 above)
Comment (jeffwashington) [reply ledger/src/blockstore_processor.rs:636]: > The validator uses these `ProcessOptions` in `core/src/validator.rs::load_blockstore()`, which calls `bank_fork_utils::load_bank_forks()`. I haven't done the plumbing here yet, but hoping it works out ok.

My question is what gets set when validator cli says to boot from local state? Shouldn't it be the same process options or shouldn't they share the same trigger mechanism?

**Review by apfitzge — APPROVED** [review]
doc-comment update looks fine, re-approve

File: ledger/src/blockstore_processor.rs (hunk 1 above)
Comment (jeffwashington) [reply ledger/src/blockstore_processor.rs:636]: ah, but that option isn't even in master yet... I guess it was never merged.

Comment (jeffwashington) [reply ledger/src/blockstore_processor.rs:636]: I assumed it was already in.

**Pull request merged**